Error reading NetCDF-4 data files with String variable type

When reading these NetCDF-4 files provided by NOAA:

http://www.star.nesdis.noaa.gov/sod/sst/iquam/v2/data.html

I get an error in the 4.6.5 toolsUI GUI about the dataset length.  When I
remove the "platform_id" variable (which is of type String, see attached
dump.cdl file) using ncdump/ncgen, the file can be read with no error.
When I try to read with version 4.5 of the toolsUI GUI, I can open the
file but get an error when reading the "platform_id" variable data.

I've written a test program (Inspect.java attached) which exhibits
the error.  When you comment back in the line that skips reading the
"platform_id" variable, the program runs fine.

Is this a design limitation of the Java NetCDF library?

Peter

[dump.cdl.txt](https://github.com/Unidata/thredds/files/312474/dump.cdl.txt)
[Inspect.java.txt](https://github.com/Unidata/thredds/files/312475/Inspect.java.txt)

Here's the error I get in my test program using toolsUI 4.6.5:

Exception in thread "main" java.io.IOException: java.io.EOFException: Reading 201604-STAR-L2i_GHRSST-SST-iQuam-V2.00-v01.0-fv01.0.nc at 59014719 file length = 59014719
	at ucar.nc2.NetcdfFile.open(NetcdfFile.java:427)
	at ucar.nc2.dataset.NetcdfDataset.openOrAcquireFile(NetcdfDataset.java:748)
	at ucar.nc2.dataset.NetcdfDataset.openDataset(NetcdfDataset.java:427)
	at ucar.nc2.dataset.NetcdfDataset.openDataset(NetcdfDataset.java:410)
	at ucar.nc2.dataset.NetcdfDataset.openDataset(NetcdfDataset.java:395)
	at ucar.nc2.dataset.NetcdfDataset.openDataset(NetcdfDataset.java:382)
	at Inspect.main(Inspect.java:8)
Caused by: java.io.EOFException: Reading 201604-STAR-L2i_GHRSST-SST-iQuam-V2.00-v01.0-fv01.0.nc at 59014719 file length = 59014719
	at ucar.unidata.io.RandomAccessFile.readFully(RandomAccessFile.java:850)
	at ucar.unidata.io.RandomAccessFile.readFully(RandomAccessFile.java:829)
	at ucar.unidata.io.RandomAccessFile.readString(RandomAccessFile.java:1520)
	at ucar.nc2.iosp.hdf5.H5header.readStringFixedLength(H5header.java:4709)
	at ucar.nc2.iosp.hdf5.H5header.readHeapString(H5header.java:4371)
	at ucar.nc2.iosp.hdf5.H5header.readAttributeData(H5header.java:1050)
	at ucar.nc2.iosp.hdf5.H5header.makeAttribute(H5header.java:946)
	at ucar.nc2.iosp.hdf5.H5header.makeAttributes(H5header.java:922)
	at ucar.nc2.iosp.hdf5.H5header.makeVariable(H5header.java:1398)
	at ucar.nc2.iosp.hdf5.H5header.makeNetcdfGroup(H5header.java:488)
	at ucar.nc2.iosp.hdf5.H5header.read(H5header.java:218)
	at ucar.nc2.iosp.hdf5.H5iosp.open(H5iosp.java:127)
	at ucar.nc2.NetcdfFile.<init>(NetcdfFile.java:1549)
	at ucar.nc2.NetcdfFile.open(NetcdfFile.java:824)
	at ucar.nc2.NetcdfFile.open(NetcdfFile.java:424)
	... 6 more

And using toolsUI 4.5 I get this:

Exception in thread "main" java.lang.IllegalStateException
	at ucar.nc2.iosp.IospHelper.readData(IospHelper.java:396)
	at ucar.nc2.iosp.IospHelper.readDataFill(IospHelper.java:292)
	at ucar.nc2.iosp.hdf5.H5iosp.readData(H5iosp.java:165)
	at ucar.nc2.iosp.hdf5.H5iosp.readData(H5iosp.java:143)
	at ucar.nc2.NetcdfFile.readData(NetcdfFile.java:1986)
	at ucar.nc2.Variable.reallyRead(Variable.java:899)
	at ucar.nc2.Variable._read(Variable.java:884)
	at ucar.nc2.Variable.read(Variable.java:695)
	at ucar.nc2.dataset.VariableDS.reallyRead(VariableDS.java:557)
	at ucar.nc2.dataset.VariableDS._read(VariableDS.java:537)
	at ucar.nc2.Variable.read(Variable.java:695)
	at ucar.nc2.Variable.read(Variable.java:641)
	at Inspect.main(Inspect.java:14)

--- a/cdm/src/main/java/ucar/nc2/iosp/hdf5/H5header.java
+++ b/cdm/src/main/java/ucar/nc2/iosp/hdf5/H5header.java
@@ -4360,6 +4360,7 @@ public class H5header {
     H5header.HeapIdentifier heapId = new HeapIdentifier(heapIdAddress);
     H5header.GlobalHeap.HeapObject ho = heapId.getHeapObject();
     if (ho == null) throw new IllegalStateException("Cant find Heap Object,heapId="+heapId);
+    if (ho.dataSize > 1000 * 1000) return String.format("Bad HeapObject.dataSize=%s", ho);
     raf.seek(ho.dataPos);
     return readStringFixedLength((int) ho.dataSize);
   }
@@ -4543,10 +4544,11 @@ public class H5header {
         int dsize = ((int) o.dataSize) + padding((int) o.dataSize, 8);
         countBytes += dsize + 16;
 
-        //System.out.printf("%d heapId=%d dataSize=%d countBytes=%d%n", count, o.id, o.dataSize, countBytes);
+        // System.out.printf("%d heapId=%d dataSize=%d countBytes=%d%n", count, o.id, o.dataSize, countBytes);
         if (o.id == 0) break; // ?? look
         if (o.dataSize < 0) break; // ran off the end, must be done
         if (countBytes < 0) break; // ran off the end, must be done
+        if (countBytes > sizeBytes) break; // ran off the end
 
         if (debugDetail)
           debugOut.println("   HeapObject  position=" + startPos + " id=" + o.id + " refCount= " + o.refCount +

--- a/cdm/src/main/java/ucar/unidata/util/GaussianLatitudes.java
+++ b/cdm/src/main/java/ucar/unidata/util/GaussianLatitudes.java
@@ -241,11 +241,12 @@ public class GaussianLatitudes {
 
   // should match http://dss.ucar.edu/datasets/common/ecmwf/ERA40/docs/std-transformations/dss_code_glwp.html
   public static void main(String args[]) {
-    int nlats = 512;
+    int nlats = 94;
     GaussianLatitudes glat = new GaussianLatitudes(nlats);
-    for (int i=0; i<nlats-1; i++) {
+    for (int i=0; i<nlats; i++) {
       System.out.print(" lat "+i+" = "+ Format.dfrac( glat.latd[i], 4));
-      System.out.print(" diff "+i+" = "+ (glat.latd[i+1]-glat.latd[i]));
+      if (i < nlats - 1)
+        System.out.print(" diff = " + (glat.latd[i + 1] - glat.latd[i]));
       System.out.println(" weight= "+ Format.dfrac( glat.gaussw[i], 6));
     }
   }

